### PR TITLE
AMQP-616: Support Direct ReplyTo in Async Template

### PR DESCRIPTION
**Pull request by garyrussell**
JIRA: https://jira.spring.io/browse/AMQP-616

As an alternative to using a SMLC reply-container for the async template,
instead, use a specially configured DMLC to listen on the channels amq.rabbitmq.reply-to
pseudo queue.

Cache these containers for reuse.

When retrieving a container from the cache, check that it's still valid before using.

### CODE REVIEW

File: spring-amqp/src/main/java/org/springframework/amqp/core/AsyncAmqpTemplate.java [hunk 1]
```diff
@@ -115,12 +115,12 @@
...
 	 * @param exchange the exchange
 	 * @param routingKey the routing key.
-	 * @param message the message.
+	 * @param object the object to convert.
 	 * @param messagePostProcessor the post processor.
 	 * @param <C> the expected result type.
 	 * @return the {@link AmqpConverterFuture}.
 	 */
-	<C> AmqpConverterFuture<C> convertSendAndReceive(String exchange, String routingKey, Object message,
+	<C> AmqpConverterFuture<C> convertSendAndReceive(String exchange, String routingKey, Object object,
```
Comment (artembilan) [inline spring-amqp/src/main/java/org/springframework/amqp/core/AsyncAmqpTemplate.java]: Recall me, please, why do we need these two `ListenableFuture` extensions without any customization and so on.
I mean `AmqpConverterFuture`  and `AmqpMessageFuture`
Thanks

File: spring-rabbit/src/main/java/org/springframework/amqp/rabbit/AsyncRabbitTemplate.java [hunk 2]
```diff
@@ -186,6 +200,35 @@ public AsyncRabbitTemplate(RabbitTemplate template, AbstractMessageListenerConta
 	}
 
 	/**
+	 * Construct an instance using the provided arguments. "Direct replyTo" is used for
+	 * replies.
+	 * @param connectionFactory the connection factory.
+	 * @param exchange the default exchange to which requests will be sent.
+	 * @param routingKey the default routing key.
```
Comment (artembilan) [inline spring-rabbit/src/main/java/org/springframework/amqp/rabbit/AsyncRabbitTemplate.java]: `@since 2.0`

File: spring-rabbit/src/main/java/org/springframework/amqp/rabbit/AsyncRabbitTemplate.java [hunk 3]
```diff
@@ -186,6 +200,35 @@ public AsyncRabbitTemplate(RabbitTemplate template, AbstractMessageListenerConta
...
+		this.template.setExchange(exchange == null ? "" : exchange);
+		this.template.setRoutingKey(routingKey);
+		this.container = null;
+		this.replyAddress = null;
+	}
+
+	/**
+	 * Construct an instance using the provided arguments. "Direct replyTo" is used for
+	 * replies.
+	 * @param template a {@link RabbitTemplate}
```
Comment (artembilan) [inline spring-rabbit/src/main/java/org/springframework/amqp/rabbit/AsyncRabbitTemplate.java]: DITO

File: spring-rabbit/src/main/java/org/springframework/amqp/rabbit/AsyncRabbitTemplate.java [hunk 4]
```diff
@@ -312,105 +352,132 @@ public RabbitMessageFuture sendAndReceive(String exchange, String routingKey, Me
...
+			}
+			channel = container.getChannel();
+		}
+		final DirectReplyToMessageListenerContainer theContainer = container;
+		container.setMessageListener(message -> {
+			onMessage(message);
+			try {
+				synchronized (this.directConsumersMonitor) {
+					if (this.running) {
+						this.directConsumers.put(theContainer);
```
Comment (artembilan) [inline spring-rabbit/src/main/java/org/springframework/amqp/rabbit/AsyncRabbitTemplate.java]: Why can't we live with he single `DirectReplyToMessageListenerContainer` instance?
Why do we need pool of them?
Thanks

File: spring-rabbit/src/test/java/org/springframework/amqp/rabbit/AsyncRabbitTemplateTests.java [hunk 5]
```diff
@@ -336,12 +420,12 @@ public ConnectionFactory connectionFactory() {
...
 		@Bean
 		public Queue requests() {
-			return new AnonymousQueue();
+			return new AnonymousQueue(new AnonymousQueue.Base64UrlNamingStrategy());
 		}
 
 		@Bean
 		public Queue replies() {
-			return new AnonymousQueue();
+			return new AnonymousQueue(new AnonymousQueue.Base64UrlNamingStrategy());
```
Comment (artembilan) [inline spring-rabbit/src/test/java/org/springframework/amqp/rabbit/AsyncRabbitTemplateTests.java]: I think we can revert this change already since `AnonymousQueue` is now Base64 by default.

File: spring-rabbit/src/main/java/org/springframework/amqp/rabbit/AsyncRabbitTemplate.java (hunk 4 above)
Comment (garyrussell) [inline spring-rabbit/src/main/java/org/springframework/amqp/rabbit/AsyncRabbitTemplate.java]: Because we need a pool of channels - a channel needs to be dedicated for a send/receive operation. I suppose we could implement the pool within the DMLC and use `consumersPerQueue` to increase the pool size I'll take a look at it.

Comment (artembilan) [inline spring-rabbit/src/main/java/org/springframework/amqp/rabbit/AsyncRabbitTemplate.java]: M-m-m. Let me guess: with the exclusive `channel` we pursue correlation.
One send and only after that one receive.

Comment (artembilan) [inline spring-rabbit/src/main/java/org/springframework/amqp/rabbit/AsyncRabbitTemplate.java]: Wow! Good to know.
Thanks

**Comment by artembilan** [comment]
Travis: 
```
:spring-rabbit:javadoc/home/travis/build/spring-projects/spring-amqp/spring-rabbit/src/main/java/org/springframework/amqp/rabbit/listener/DirectMessageListenerContainer.java:591: error: @param name not found

	 * @param the consumer.
	          ^

/home/travis/build/spring-projects/spring-amqp/spring-rabbit/src/main/java/org/springframework/amqp/rabbit/listener/DirectMessageListenerContainer.java:593: warning: no @param for consumer

	protected void consumerRemoved(SimpleConsumer consumer) {
	               ^
```

Will fix on merge if only concern.

File: spring-rabbit/src/main/java/org/springframework/amqp/rabbit/listener/DirectReplyToMessageListenerContainer.java [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+			Channel channel = null;
+			while (channel == null) {
+				for (SimpleConsumer consumer : getConsumers()) {
+					if (this.inUseConsumerChannels.putIfAbsent(consumer.getChannel(), Boolean.TRUE) == null) {
+						channel = consumer.getChannel();
+						break;
+					}
+				}
+				if (channel == null) {
+					this.consumerCount++;
```
Comment (artembilan) [inline spring-rabbit/src/main/java/org/springframework/amqp/rabbit/listener/DirectReplyToMessageListenerContainer.java]: I think we are missing to decrease this variable when we remove consumers

Comment (garyrussell) [inline spring-rabbit/src/main/java/org/springframework/amqp/rabbit/listener/DirectReplyToMessageListenerContainer.java]: The count is never decremented; we only remove them from the in use map - the monitor thread will recover any consumers that die.

**Comment by garyrussell** [comment]
All good comments; thanks; working on a new commit - almost ready but will finish tomorrow.

File: spring-rabbit/src/main/java/org/springframework/amqp/rabbit/listener/DirectReplyToMessageListenerContainer.java (hunk 6 above)
Comment (artembilan) [inline spring-rabbit/src/main/java/org/springframework/amqp/rabbit/listener/DirectReplyToMessageListenerContainer.java]: Well, but do we need extra consumers, if the pace from produces is calming down?

File: spring-rabbit/src/main/java/org/springframework/amqp/rabbit/listener/DirectMessageListenerContainer.java [hunk 7]
```diff
@@ -447,8 +454,12 @@ private void checkMissingQueues(String[] queueNames) {
 	}
 
 	private void consumeFromQueue(String queue) {
-		for (int i = 0; i < this.consumersPerQueue; i++) {
-			doConsumeFromQueue(queue);
+		List<SimpleConsumer> list = this.consumersByQueue.get(queue);
+		// Possible race with setConsumersPerQueue and the task launched by start()
+		if (list == null || list.size() == 0) {
```
Comment (artembilan) [inline spring-rabbit/src/main/java/org/springframework/amqp/rabbit/listener/DirectMessageListenerContainer.java]: `CollectionUtils.isEmpty()`, but, eh..., minor.

File: spring-rabbit/src/main/java/org/springframework/amqp/rabbit/listener/DirectReplyToMessageListenerContainer.java [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+	 * Set cancelConsumer to true if the client is not prepared to handle/discard a
+	 * late arriving reply.
+	 * @param channel the channel.
+	 * @param cancelConsumer true to cancel the consumer.
+	 * @param message a message to be included in the cancel event if cancelConsumer is true.
+	 */
+	public void releaseConsumerFor(Channel channel, boolean cancelConsumer, String message) {
+		SimpleConsumer consumer = this.inUseConsumerChannels.remove(channel);
+		if (consumer != null && cancelConsumer) {
+			this.logger.error("Consumer canceled by client" + consumer);
```
Comment (artembilan) [inline spring-rabbit/src/main/java/org/springframework/amqp/rabbit/listener/DirectReplyToMessageListenerContainer.java]: What is the purpose of this `error()`? Why simple `debug()` is not enough?
and at least whitespace after "client" word.

File: spring-rabbit/src/main/java/org/springframework/amqp/rabbit/listener/DirectReplyToMessageListenerContainer.java [hunk 9]
```diff
@@ -0,0 +1,198 @@
...
+			consumer.cancelConsumer("Consumer " + this + " canceled due to " + message);
```
Comment (artembilan) [inline spring-rabbit/src/main/java/org/springframework/amqp/rabbit/listener/DirectReplyToMessageListenerContainer.java]: Or `message` must never be `null`, or we should reconsider this sentence to honor `null` variant.

Comment (garyrussell) [inline spring-rabbit/src/main/java/org/springframework/amqp/rabbit/listener/DirectReplyToMessageListenerContainer.java]: I don't understand these comments - the `message` is only used when `cancelConsumer` is true. What am I missing? Please elaborate.

Comment (artembilan) [inline spring-rabbit/src/main/java/org/springframework/amqp/rabbit/listener/DirectReplyToMessageListenerContainer.java]: Ah! Indeed.
So, we should ensure that `message` isn't `null`, when `cancelConsumer == true`.
I don't say that only one `cancelMessage` and check it for `null` would be good solution (although for computer it is really enough :smile: ), but at least we will provide the proper message for human being. Otherwise we can end up with the:
```
... canceled due to null
```

**Comment by artembilan** [comment]
Merged as https://github.com/spring-projects/spring-amqp/commit/8cdbd4cd170a620ec23f7901ad570859c5011b22
